File: 1.HTTP-AND-CORS/code/typescript/idempotency_in_practice_4.ts
import express, { Request, Response } from 'express';

const app = express();

// key -> string result. Use Redis with a TTL in production.
const seen = new Map<string, string>();

function charge(req: Request): string {
    return "payment_success";
}

app.post('/charge', (req: Request, res: Response) => {
    const key = req.header('Idempotency-Key');
    if (!key) {
        res.status(400).send('missing Idempotency-Key');
        return;
    }
    
    if (seen.has(key)) {                      // replay: return the stored result
        const cached = seen.get(key);
        res.status(200).send(cached);
        return;
    }
    
    const result = charge(req);               // the real, non-idempotent work
    seen.set(key, result);                    // remember it BEFORE responding
    
    res.status(201).send(result);
});

File: 1.HTTP-AND-CORS/code/rust/idempotency_in_practice_4.rs
use actix_web::{web, HttpRequest, HttpResponse, Responder};
use std::collections::HashMap;
use std::sync::Mutex;

// key -> String result. Use Redis with a TTL in production.
struct AppState {
    seen: Mutex<HashMap<String, String>>,
}

fn charge(_req: &HttpRequest) -> String {
    "payment_success".to_string()
}

async fn create_payment(req: HttpRequest, data: web::Data<AppState>) -> impl Responder {
    let key_header = req.headers().get("Idempotency-Key");
    
    let key = match key_header {
        Some(k) => match k.to_str() {
            Ok(s) => s.to_string(),
            Err(_) => return HttpResponse::BadRequest().body("invalid Idempotency-Key"),
        },
        None => return HttpResponse::BadRequest().body("missing Idempotency-Key"),
    };
    
    let mut seen = data.seen.lock().unwrap();
    if let Some(cached) = seen.get(&key) {    // replay: return the stored result
        return HttpResponse::Ok().body(cached.clone());
    }
    
    let result = charge(&req);                // the real, non-idempotent work
    seen.insert(key, result.clone());         // remember it BEFORE responding
    
    HttpResponse::Created().body(result)
}
